Modules/Authentication/Roles/RoleInterceptor.cs

```diff
@@ -0,0 +1,61 @@
+﻿using GenHTTP.Api.Content;
+using GenHTTP.Api.Content.Authentication;
+using GenHTTP.Api.Protocol;
+
+using GenHTTP.Modules.Reflection;
+using GenHTTP.Modules.Reflection.Operations;
+
+namespace GenHTTP.Modules.Authentication.Roles;
+
+public class RoleInterceptor : IOperationInterceptor
+{
+    private string[]? _Roles;
+
+    public void Configure(object attribute)
+    {
+        if (attribute is RequireRoleAttribute roleAttribute)
+        {
+            _Roles = roleAttribute.Roles;
+        }
+    }
+
+    public ValueTask<InterceptionResult?> InterceptAsync(IRequest request, Operation operation, IReadOnlyDictionary<string, object?> arguments)
+    {
+        if (_Roles?.Length > 0)
+        {
+            var user = request.GetUser<IUser>();
+
+            if (user == null)
+            {
+                throw new ProviderException(ResponseStatus.Unauthorized, "Credentials are required for this endpoint");

```

Comment:
Term `Credentials` could be perhaps a bit misleading in this context as the authorization method may vary. I'd either change it to `Authorization is required ...` or `User is not authorized.`.